windows/lcow_runtime.go

```diff
@@ -0,0 +1,400 @@
+// +build windows
+
+package windows
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/Microsoft/hcsshim"
+	"github.com/boltdb/bolt"
+	eventsapi "github.com/containerd/containerd/api/services/events/v1"
+	containerdtypes "github.com/containerd/containerd/api/types"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/events"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/runtime"
+	"github.com/containerd/containerd/typeurl"
+	"github.com/containerd/containerd/windows/hcsshimopts"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+const (
+	lcowRuntimeName = "lcow"
+)
+
+var (
+	lcowPluginID = fmt.Sprintf("%s.%s", plugin.RuntimePlugin, lcowRuntimeName)
+)
+
+var _ = (runtime.Runtime)(&LCOWRuntime{})
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		ID:   lcowRuntimeName,
+		Type: plugin.RuntimePlugin,
+		Init: NewLCOWPlugin,
+		Requires: []plugin.PluginType{
+			plugin.MetadataPlugin,
+		},
+	})
+}
+
+func NewLCOWPlugin(ic *plugin.InitContext) (interface{}, error) {
+	if err := os.MkdirAll(ic.Root, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create state directory at %s", ic.Root)
+	}
+
+	m, err := ic.Get(plugin.MetadataPlugin)
+	if err != nil {
+		return nil, err
+	}
+
+	cd := filepath.Join(ic.Root, cacheDirectory)
+	sd := filepath.Join(ic.Root, scratchDirectory)
+
+	// Make sure the scratch directory is created under dataRoot
+	if err := os.MkdirAll(sd, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create scratch directory at %s", ic.Root)
+	}
+
+	// Make sure the cache directory is created under dataRoot
+	if err := os.MkdirAll(cd, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create cache directory at %s", ic.Root)
+	}
+
+	r := &LCOWRuntime{
+		root:    ic.Root,
+		pidPool: newPidPool(),
+
+		events:    make(chan interface{}, 4096),
+		publisher: ic.Events,
+		// TODO(mlaventure): lcow needs a stat monitor
+		monitor: nil,
+		tasks:   runtime.NewTaskList(),
+		db:      m.(*bolt.DB),
+
+		cachedSandboxFile: filepath.Join(cd, sandboxFilename),
+		cachedScratchFile: filepath.Join(cd, scratchFilename),
+		cache:             make(map[string]*cacheItem),
+		serviceVms:        make(map[string]*serviceVMItem),
+		globalMode:        false,
+	}
+
+	// Load our existing containers and kill/delete them. We don't support
+	// reattaching to them
+	r.cleanup(ic.Context)
+
+	return r, nil
+}
+
+type LCOWRuntime struct {
+	sync.Mutex
+
+	root    string
+	pidPool *pidPool
+
+	publisher events.Publisher
+	events    chan interface{}
+
+	monitor runtime.TaskMonitor
+	tasks   *runtime.TaskList
+	db      *bolt.DB
+
+	cachedSandboxFile  string                    // Location of the local default-sized cached sandbox.
```

Comment:
I agree that the name makes more sense, but for ease of transition to the snapshotter and keeping in sync with the similar code in moby/moby (which will eventually end up in containerd), I'd like to keep this name for now.

Reply:
Will only yield there if the name get changed to something proper once the snapshotter is introduced. I'd rather not have confusing code in containerd just because we had to port over code from docker/moby. We want new contributor to easily be able to understand the code as much as possible without our help (for our own benefit :))

Reply:
Absolutely. I do not intend to keep a confusing name for legacy sake. It'll just make my life easier to temporarily keep these in sync during the transition.